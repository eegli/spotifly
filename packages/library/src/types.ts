type ExportedLibrary = 'full' | 'light';

export type Options = {
  token: string;
  type?: ExportedLibrary;
  genres?: boolean;
  features?: boolean;
  compact?: boolean;
  outDir?: string;
  since?: string;
  last?: number;
};

export type TrackLight = {
  id: string;
  name: string;
  album: {
    name: string;
    id: string;
  };
  artists: {
    id: string;
    name: string;
  }[];
};

export type TrackFull = SpotifyApi.TrackObjectFull;

type AnyTrack = TrackLight | TrackFull;

type LibraryExport<T> = {
  meta: {
    date_generated: string;
    output_type: ExportedLibrary;
  };
  library: Library<T>;
};

export type Library<T = AnyTrack> = {
  added_at: string;
  track: T & {
    genres?: string[][];
    features?: SpotifyApi.AudioFeaturesObject;
  };
}[];

export type LibraryHandler = (
  opts: Options,
) => Promise<LibraryExport<AnyTrack>>;
