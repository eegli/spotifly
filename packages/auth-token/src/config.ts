import { Parser } from '@eegli/tinyparse';
import log from '@spotifly/utils/log';

export const options = new Parser()
  .setMeta({
    command: 'spotifly auth',
    help: {
      command: 'help',
      longFlag: '--help',
      shortFlag: '-h',
    },
  })
  .option('clientId', {
    required: true,
    longFlag: '--client-id',
    defaultValue: '',
    description: 'The client id of your Spotify application',
  })
  .option('clientSecret', {
    required: true,
    longFlag: '--client-secret',
    defaultValue: '',
    description: 'The client secret of your Spotify application',
  })
  .option('scopes', {
    longFlag: '--scopes',
    shortFlag: '-s',
    defaultValue: 'user-read-email',
    description:
      'Spotify authorization scopes. Multiple scopes need to be separated by a space',
  })
  .option('port', {
    longFlag: '--port',
    shortFlag: '-p',
    defaultValue: 3000,
    description: 'Port for the localhost redirect URL',
  })
  .option('outDir', {
    longFlag: '--out-dir',
    shortFlag: '-o',
    defaultValue: '',
    description: 'Custom relative output directory',
  })
  .option('fileName', {
    longFlag: '--file-name',
    shortFlag: '-f',
    defaultValue: 'spotify-token',
    description: 'Custom name for the output JSON file',
  })
  .option('noEmit', {
    longFlag: '--no-emit',
    defaultValue: false,
    description: 'When set to true, does not save the token to the file system',
  })
  .onError(({ error, usage }) => {
    log.error('Error: ' + error.message);
    log.log(usage);
  });

export type Options = typeof options;
