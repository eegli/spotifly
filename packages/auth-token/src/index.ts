export { authorize } from './authorize';
export type { Options } from './types';
