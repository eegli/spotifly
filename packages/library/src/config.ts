import { createParser } from '@eegli/tinyparse';
import { Options } from './types';

const ZERO_DATE = new Date(0);

export const defaultConfig: Required<Options> = {
  token: '',
  type: 'light',
  genres: false,
  features: false,
  compact: false,
  outDir: '',
  since: ZERO_DATE.toISOString(),
  last: Infinity,
};

export const { parse, help } = createParser(defaultConfig, {
  decamelize: true,
  options: {
    token: {
      required: true,
      description:
        'A Spotify user access token. Requires at least the scope "user-library-read"',
    },
    type: {
      description:
        "Output type per track. Either 'full' or 'light'. Default: 'light'",
      customValidator: {
        isValid(value): value is 'light' | 'full' {
          return value === 'light' || value === 'full';
        },
        errorMessage(value) {
          return `Invalid value '${value}' for option 'type'. Allowed options are 'full' and 'light'`;
        },
      },
    },
    genres: {
      description: 'Include artist genres for each track. Default: false',
    },
    features: {
      description: 'Include audio features for each track. Default: false',
    },
    since: {
      description:
        'Only include tracks added after this date. The date string must be formatted according to the ECMAScript Date Time String Format, e.g.: "YYYY-MM-DD". Default: All tracks',
      customValidator: {
        isValid(value): value is string {
          if (typeof value !== 'string') return false;
          return !isNaN(new Date(value).getTime());
        },
        errorMessage(value) {
          return `Invalid value '${value}' for option 'since'. Expected a valid date string`;
        },
      },
    },
    last: {
      description:
        'Only include the last n (most recent) tracks. Default: All tracks',
    },
    compact: {
      description:
        'Output more compact/minified JSON and save disk space. Default: false',
    },
    outDir: {
      description:
        'Custom relative output directory. Default: Current directory',
    },
  },
});
