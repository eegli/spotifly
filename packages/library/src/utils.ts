import cliProgress from 'cli-progress';

export const createProgressBar = (items: string) =>
  new cliProgress.SingleBar({
    format:
      `Fetching ${items}\t |` +
      '{bar} {percentage}% | ETA: {eta}s | {value}/{total}',
    barCompleteChar: '\u25A0',
    barIncompleteChar: ' ',
    hideCursor: true,
  });

/**
 * Checks if the first date as a string is before another date
 */
export const isBeforeDate = (
  date1: string | Date,
  date2: string | Date,
): boolean => {
  const d1 = date1 instanceof Date ? date1.getTime() : Date.parse(date1);
  const d2 = date2 instanceof Date ? date2.getTime() : Date.parse(date2);
  if (isNaN(d1) || isNaN(d2)) {
    return false;
  }
  return d1 < d2;
};
