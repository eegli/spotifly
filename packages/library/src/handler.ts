import { initialize, isError as isSpotiflyError } from '@spotifly/core';
import { writeJSON } from '@spotifly/utils/fs';
import log from '@spotifly/utils/log';
import { Result } from '@spotifly/utils/types';
import type { Library, LibraryExport, LibraryParams } from './types';
import { createProgressBar, isBeforeDate } from './utils';

export const libraryHandler = async ({
  options,
  globals,
}: LibraryParams): Promise<Result<LibraryExport>> => {
  let progress = createProgressBar('user library');
  try {
    const spotifyClient = initialize({ accessToken: globals.token });

    let library: Library = [];

    progress.start(0, 0);

    let nextPage: string | null = null;
    let offset = 0;

    fetchLoop: do {
      const { data } = await spotifyClient.Tracks.getUsersSavedTracks({
        limit: 50,
        offset,
      });
      progress.setTotal(data.total);
      progress.increment(data.items.length);

      for (let i = 0; i < data.items.length; i++) {
        const track = data.items[i];
        if (library.length === options.last) {
          break fetchLoop;
        }
        if (isBeforeDate(track.added_at, options.since)) {
          break fetchLoop;
        }
        library.push(track);
      }
      nextPage = data.next;
      offset += 50;
    } while (nextPage);

    progress.stop();

    // Reduce library if necessary
    if (options.type === 'light') {
      library = library.reduce(
        (acc, curr) => {
          acc.push({
            added_at: curr.added_at,
            track: {
              id: curr.track.id,
              name: curr.track.name,
              album: {
                id: curr.track.album.id,
                name: curr.track.album.name,
              },
              artists: curr.track.artists.map(a => ({
                id: a.id,
                name: a.name,
              })),
            },
          });
          return acc;
        },
        <Library>[],
      );
    }

    // Add genres if specified
    if (options.genres) {
      const artists = library.map(t => t.track.artists.map(a => a.id)).flat();
      const artistIds = [...new Set<string>(artists)];
      const genres: Record<string, string[]> = {};

      progress = createProgressBar('artists');
      progress.start(artistIds.length, 0);

      await spotifyClient.Artists.getAllArtists(artistIds)(({ data }) => {
        data.artists.forEach(artist => {
          genres[artist.id] = artist.genres;
        });
        progress.increment(data.artists.length);
      });

      progress.stop();
      library.forEach(({ track }) => {
        track.genres = track.artists.map(({ id }) => genres[id]);
      });
    }

    // Add audio features if specified
    if (options.features) {
      const trackIds = library.map(t => t.track.id);
      const features: Record<string, SpotifyApi.AudioFeaturesObject> = {};

      progress = createProgressBar('audio features');

      progress.start(trackIds.length, 0);

      await spotifyClient.Tracks.getAllAudioFeatures(trackIds)(({ data }) => {
        data.audio_features.forEach(f => {
          features[f.id] = f;
        });
        progress.increment(data.audio_features.length);
      });

      progress.stop();
      library.forEach(({ track }) => {
        track.features = features[track.id];
      });
    }

    const libExport: LibraryExport = {
      meta: {
        date_generated: new Date().toISOString(),
      },
      library,
    };
    const fileName = 'spotify-library';
    const outFile = await writeJSON({
      fileName,
      path: options.outDir,
      data: libExport,
      compact: options.compact,
    });

    log.info(`Success! Library written to ${outFile}`);
    return {
      success: true,
      value: libExport,
    };
  } catch (error) {
    progress.stop();
    if (isSpotiflyError(error) && error.response) {
      const { status, message } = error.response.data.error;
      return {
        success: false,
        error: `Failed to call Spotify API: Status ${status}, ${message}`,
      };
    } else if (error instanceof Error) {
      return {
        success: false,
        error: error.message,
      };
    } else {
      return {
        success: false,
        error: 'Something went wrong',
      };
    }
  }
};
