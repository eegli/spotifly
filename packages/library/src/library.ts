import { writeJSON } from '@spotifly/utils';
import axios from 'axios';
import { defaultConfig } from './config';
import type {
  Library,
  LibraryExport,
  MultipleArtists,
  MultipleAudioFeatures,
  Options,
  TrackFull,
  TrackLight,
  UsersSavedTracks,
} from './types';
import { chunkify, createProgressBar } from './utils';

export const getLibrary = async (
  options: Options
): Promise<LibraryExport<TrackLight | TrackFull>> => {
  try {
    const config = { ...defaultConfig, ...options };
    const lib: LibraryExport<TrackFull | TrackLight> = {
      meta: {
        date_generated: new Date().toISOString(),
        output_type: config.type,
      },
      library: [],
    };
    let progress = createProgressBar('user library');

    const pagination = {
      nextUrl: <string | null>'https://api.spotify.com/v1/me/tracks',
      start: 0,
      end: 50,
    };

    progress.start(0, 0);
    while (pagination.nextUrl) {
      const { data }: UsersSavedTracks = await axios.get(pagination.nextUrl, {
        headers: {
          Authorization: `Bearer ${config.token}`,
        },
        params: {
          limit: 50,
        },
      });

      progress.setTotal(data.total);
      progress.increment(data.items.length);

      data.items.forEach(el => {
        lib.library.push(el);
      });

      pagination.nextUrl = data.next;
      pagination.start = pagination.end;
      pagination.end += data.limit;
    }

    progress.stop();

    // Reduce library if necessary
    if (config.type === 'light') {
      lib.library = lib.library.reduce((acc, curr) => {
        acc.push({
          added_at: curr.added_at,
          track: {
            id: curr.track.id,
            name: curr.track.name,
            album: {
              id: curr.track.album.id,
              name: curr.track.album.name,
            },
            artists: curr.track.artists.map(a => ({
              id: a.id,
              name: a.name,
            })),
          },
        });
        return acc;
      }, <Library<TrackLight>>[]);
    }

    // Add genres if specified
    if (config.genres) {
      const artists = lib.library
        .map(t => t.track.artists.map(a => a.id))
        .flat();
      const artistIds = new Set<string>(artists);
      const genres: Record<string, string[]> = {};

      // Max 50 items per request to Spotify
      const chunkSize = 50;

      progress = createProgressBar('artists');
      progress.start(artistIds.size, 0);

      for await (const chunk of chunkify(artistIds, chunkSize)) {
        const { data }: MultipleArtists = await axios.get(
          'https://api.spotify.com/v1/artists',
          {
            headers: {
              Authorization: `Bearer ${config.token}`,
            },
            params: {
              ids: chunk.join(','),
            },
          }
        );
        data.artists.forEach(artist => {
          genres[artist.id] = artist.genres;
        });
        progress.increment(chunkSize);
      }
      progress.stop();
      lib.library.forEach(({ track }) => {
        track.genres = track.artists.map(({ id }) => genres[id]);
      });
    }

    // Add audio features if specified
    if (config.features) {
      const songIds = lib.library.map(t => t.track.id);
      const features: Record<string, SpotifyApi.AudioFeaturesObject> = {};

      const chunkSize = 50;
      progress = createProgressBar('audio features');

      progress.start(songIds.length, 0);
      for await (const chunk of chunkify(songIds, 50)) {
        const { data }: MultipleAudioFeatures = await axios.get(
          'https://api.spotify.com/v1/audio-features',
          {
            headers: {
              Authorization: `Bearer ${config.token}`,
            },
            params: {
              ids: chunk.join(','),
            },
          }
        );

        data.audio_features.forEach(f => {
          features[f.id] = f;
        });
        progress.increment(chunkSize);
      }
      progress.stop();
      lib.library.forEach(({ track }) => {
        track.features = features[track.id];
      });
    }

    const outDir = await writeJSON({
      fileName: 'spotify-library',
      path: config.outDir,
      data: lib,
      compact: config.compact,
    });
    console.info("Success! Library written to '%s'", outDir);
    return lib;
  } catch (e) {
    if (axios.isAxiosError(e)) {
      const error = new Error(e.message);
      Object.assign(error, e.toJSON());
      throw error;
    }
    throw e;
  }
};
