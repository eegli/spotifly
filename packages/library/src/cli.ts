import { help as defaultHelp, parse } from './config';
import { libraryHandler } from './handler';

export const callback = async (argv: string[]): Promise<void> => {
  const config = await parse(argv);
  await libraryHandler(config);
};

export const help = () =>
  defaultHelp({
    title: 'Command-line usage:',
  });

export { default as pkg } from '../package.json';
