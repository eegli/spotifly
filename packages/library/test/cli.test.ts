import { help, parse } from '../src/config';

describe('Library CLI', () => {
  test('invalid parsing', async () => {
    await expect(parse([])).rejects.toThrow();
    await expect(
      parse({ token: 'abc', since: '2022-12-121' })
    ).rejects.toThrow();
  });
  test('valid parsing', async () => {
    await expect(parse({ token: 'abc' })).resolves.toMatchSnapshot();
    await expect(
      parse({ token: 'abc', since: '2022-12-12' })
    ).resolves.not.toThrow();
  });
  test('help command', () => {
    expect(help()).toMatchInlineSnapshot(`
      "Usage

      Required
         --token [string]
         A Spotify user access token. Requires at least the scope \\"user-library-read\\"

      Optional
         --type [string]
         Output type per track. Either 'full' or 'light'. Default: 'light'

         --genres [boolean]
         Include artist genres for each track. Default: false

         --features [boolean]
         Include audio features for each track. Default: false

         --since [string]
         Only include tracks added after this date. The date string must be formatted according to the ECMAScript Date Time String Format, e.g.: \\"YYYY-MM-DD\\". Default: All tracks

         --last [number]
         Only include the last n (most recent) tracks. Default: All tracks

         --compact [boolean]
         Output more compact/minified JSON and save disk space. Default: false

         --outDir [string]
         Custom relative output directory. Default: Current directory"
    `);
  });
});
