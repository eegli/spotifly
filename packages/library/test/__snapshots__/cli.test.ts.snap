// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Library CLI valid parsing 1`] = `
{
  "compact": false,
  "features": false,
  "genres": false,
  "last": Infinity,
  "outDir": "",
  "since": "1970-01-01T00:00:00.000Z",
  "token": "abc",
  "type": "light",
}
`;
